source/extensions/common/aws/sigv4a_signer_impl.h

```diff
@@ -0,0 +1,133 @@
+#pragma once
+
+#include <utility>
+
+#include "source/common/common/logger.h"
+#include "source/common/common/matchers.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/headers.h"
+#include "source/common/singleton/const_singleton.h"
+#include "source/extensions/common/aws/credentials_provider.h"
+#include "source/extensions/common/aws/signer.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Aws {
+
+class SigV4ASignatureHeaderValues {
+public:
+  const Http::LowerCaseString ContentSha256{"x-amz-content-sha256"};
+  const Http::LowerCaseString Date{"x-amz-date"};
+  const Http::LowerCaseString SecurityToken{"x-amz-security-token"};
+  const Http::LowerCaseString RegionSet{"x-amz-region-set"};
+};
+
+using SigV4ASignatureHeaders = ConstSingleton<SigV4ASignatureHeaderValues>;
+
+class SigV4ASignatureConstantValues {
+public:
+  const std::string Aws4Request{"aws4_request"};
+  const std::string SigV4AAuthorizationHeaderFormat{
+      "AWS4-ECDSA-P256-SHA256 Credential={}/{}, SignedHeaders={}, Signature={}"};
+  const std::string SigV4ACredentialScopeFormat{"{}/{}/aws4_request"};
+  const std::string HashedEmptyString{
+      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"};
+  const std::string SigV4ASignatureVersion{"AWS4A"};
+  const std::string SigV4AStringToSignFormat{"AWS4-ECDSA-P256-SHA256\n{}\n{}\n{}"};
+  const std::string SigV4ALabel = "AWS4-ECDSA-P256-SHA256";
+
+  const std::string LongDateFormat{"%Y%m%dT%H%M00Z"};
+  const std::string ShortDateFormat{"%Y%m%d"};
+  const std::string UnsignedPayload{"UNSIGNED-PAYLOAD"};
+};
+
+enum SigV4AKeyDerivationResult {
+  AkdrSuccess,
+  AkdrNextCounter,
+  AkdrFailure,
+};
+
+using SigV4ASignatureConstants = ConstSingleton<SigV4ASignatureConstantValues>;
+
+using AwsSigningHeaderExclusionVector = std::vector<envoy::type::matcher::v3::StringMatcher>;
+
+/**
+ * Implementation of the Signature V4A signing process.
+ * See https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+ */
+
+class SigV4ASignerImpl : public Signer, public Logger::Loggable<Logger::Id::aws> {
```

Comment:
is it fair to say some of the functions are the same between the SigV4SignerImpl & SigV4aSignerImpl?

If so can we create a parent class to hold all the common functionalities so that we don't accidentally introduce a bug in future with less repeated code?

Reply:
this is a fair comment. i will look to update this to incorporate common code across both